src/prefect/task_runs.py

```diff
@@ -195,6 +198,26 @@ async def wait_for_task_run(
                 # Remove the event from the cache after it has been waited on
                 instance._completion_events.pop(task_run_id, None)
 
+    @classmethod
+    def add_done_callback(cls, task_run_id: uuid.UUID, callback):
+        """
+        Add a callback to be called when a task run finishes.
+
+        Args:
+            task_run_id: The ID of the task run to wait for.
+            callback: The callback to call when the task run finishes.
+        """
+        instance = cls.instance()
+        with instance._observed_completed_task_runs_lock:
+            if task_run_id in instance._observed_completed_task_runs:
+                callback()
+                return
+
+        with instance._completion_events_lock:
+            # Cache the event for the task run ID so the consumer can set it
+            # when the event is received
+            instance._completion_events[task_run_id] = callback
```

Comment:
this doesn't fit the typing on `instance._completion_events` - is this intentional?